com.hazelcast.jet.cdc.mysql.MySqlCdcNetworkIntegrationTest.when_databaseShutdownDuringBinlogReading[fail]

_master_ (commit 7132dae97adf8b2b19df4676f9da5e6e04a6831e)

Failed on Sonar nightly build: http://jenkins.hazelcast.com/job/jet-oss-master-sonar-nightly/622/testReport/junit/com.hazelcast.jet.cdc.mysql/MySqlCdcNetworkIntegrationTest/when_databaseShutdownDuringBinlogReading_fail_/

Stacktrace:
```
org.testcontainers.containers.ContainerLaunchException: Container startup failed
	at org.testcontainers.containers.GenericContainer.doStart(GenericContainer.java:330)
	at org.testcontainers.containers.GenericContainer.start(GenericContainer.java:311)
	at com.hazelcast.jet.cdc.mysql.MySqlCdcNetworkIntegrationTest.initMySql(MySqlCdcNetworkIntegrationTest.java:355)
	at com.hazelcast.jet.cdc.mysql.MySqlCdcNetworkIntegrationTest.when_databaseShutdownDuringBinlogReading(MySqlCdcNetworkIntegrationTest.java:258)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:114)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:106)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: org.rnorth.ducttape.RetryCountExceededException: Retry limit hit with exception
	at org.rnorth.ducttape.unreliables.Unreliables.retryUntilSuccess(Unreliables.java:88)
	at org.testcontainers.containers.GenericContainer.doStart(GenericContainer.java:323)
	... 15 more
Caused by: org.testcontainers.containers.ContainerLaunchException: Could not create/start container
	at org.testcontainers.containers.GenericContainer.tryStart(GenericContainer.java:498)
	at org.testcontainers.containers.GenericContainer.lambda$doStart$0(GenericContainer.java:325)
	at org.rnorth.ducttape.unreliables.Unreliables.retryUntilSuccess(Unreliables.java:81)
	... 16 more
Caused by: com.github.dockerjava.api.exception.InternalServerErrorException: {"message":"driver failed programming external connectivity on endpoint reverent_kapitsa (9f4ec07388a040814fdf1f40c9dc328bdfe395b01f431acd1ce975863144a471): Error starting userland proxy: listen tcp 0.0.0.0:40708: bind: address already in use"}

	at com.github.dockerjava.okhttp.OkHttpInvocationBuilder.execute(OkHttpInvocationBuilder.java:293)
	at com.github.dockerjava.okhttp.OkHttpInvocationBuilder.execute(OkHttpInvocationBuilder.java:271)
	at com.github.dockerjava.okhttp.OkHttpInvocationBuilder.post(OkHttpInvocationBuilder.java:116)
	at com.github.dockerjava.core.exec.StartContainerCmdExec.execute(StartContainerCmdExec.java:31)
	at com.github.dockerjava.core.exec.StartContainerCmdExec.execute(StartContainerCmdExec.java:13)
	at com.github.dockerjava.core.exec.AbstrSyncDockerCmdExec.exec(AbstrSyncDockerCmdExec.java:21)
	at com.github.dockerjava.core.command.AbstrDockerCmd.exec(AbstrDockerCmd.java:35)
	at com.github.dockerjava.core.command.StartContainerCmdImpl.exec(StartContainerCmdImpl.java:43)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.testcontainers.dockerclient.AuditLoggingDockerClient.lambda$wrappedCommand$14(AuditLoggingDockerClient.java:102)
	at com.sun.proxy.$Proxy45.exec(Unknown Source)
	at org.testcontainers.containers.GenericContainer.tryStart(GenericContainer.java:419)
	... 18 more
```

Standard output:
```
2020-09-20 13:40:31,368 [ INFO] [when_databaseShutdownDuringBinlogReading] [o.t.d.DockerClientProviderStrategy]: Loaded org.testcontainers.dockerclient.EnvironmentAndSystemPropertyClientProviderStrategy from ~/.testcontainers.properties, will try it first
2020-09-20 13:40:32,130 [ INFO] [when_databaseShutdownDuringBinlogReading] [o.t.d.EnvironmentAndSystemPropertyClientProviderStrategy]: Found docker client settings from environment
2020-09-20 13:40:32,131 [ INFO] [when_databaseShutdownDuringBinlogReading] [o.t.d.DockerClientProviderStrategy]: Found Docker environment with Environment variables, system properties and defaults. Resolved dockerHost=unix:///var/run/docker.sock
2020-09-20 13:40:32,362 [ INFO] [when_databaseShutdownDuringBinlogReading] [o.t.DockerClientFactory]: Docker host IP address is 172.17.0.1
2020-09-20 13:40:32,394 [ INFO] [when_databaseShutdownDuringBinlogReading] [o.t.DockerClientFactory]: Connected to docker: 
  Server Version: 18.06.1-ce
  API Version: 1.38
  Operating System: Red Hat Enterprise Linux
  Total Memory: 386667 MB
2020-09-20 13:40:32,536 [ WARN] [when_databaseShutdownDuringBinlogReading] [o.t.u.RegistryAuthLocator]: Failure when attempting to lookup auth config (dockerImageName: testcontainersofficial/ryuk:0.3.0, configFile: /home/jenkins/.docker/config.json. Falling back to docker-java default behaviour. Exception message: /home/jenkins/.docker/config.json (No such file or directory)
2020-09-20 13:40:33,659 [ INFO] [when_databaseShutdownDuringBinlogReading] [o.t.DockerClientFactory]: Ryuk started - will monitor and terminate Testcontainers containers on JVM exit
2020-09-20 13:40:33,659 [ INFO] [when_databaseShutdownDuringBinlogReading] [o.t.DockerClientFactory]: Checking the system...
2020-09-20 13:40:33,660 [ INFO] [when_databaseShutdownDuringBinlogReading] [o.t.DockerClientFactory]: ?? Docker server version should be at least 1.6.0
2020-09-20 13:40:33,752 [ INFO] [when_databaseShutdownDuringBinlogReading] [o.t.DockerClientFactory]: ?? Docker environment should have more than 2GB free disk space
2020-09-20 13:40:33,897 [ INFO] [when_databaseShutdownDuringBinlogReading] [d.2]]: Creating container for image: debezium/example-mysql:1.2
2020-09-20 13:40:34,632 [ INFO] [when_databaseShutdownDuringBinlogReading] [d.2]]: Starting container with ID: a09f6c19aae64ebfe1ba03092734383cc2acd1528cdc84c3ff5b09ac72a25782
2020-09-20 13:40:34,865 [ERROR] [when_databaseShutdownDuringBinlogReading] [d.2]]: Could not start container
com.github.dockerjava.api.exception.InternalServerErrorException: {"message":"driver failed programming external connectivity on endpoint gracious_kapitsa (d426ea32b4b02d40070c3aea6261cd6d42820d18242aa599ab1f6e4ae7b616e8): Error starting userland proxy: listen tcp 0.0.0.0:40708: bind: address already in use"}

	at com.github.dockerjava.okhttp.OkHttpInvocationBuilder.execute(OkHttpInvocationBuilder.java:293) ~[testcontainers-1.14.3.jar:?]
	at com.github.dockerjava.okhttp.OkHttpInvocationBuilder.execute(OkHttpInvocationBuilder.java:271) ~[testcontainers-1.14.3.jar:?]
	at com.github.dockerjava.okhttp.OkHttpInvocationBuilder.post(OkHttpInvocationBuilder.java:116) ~[testcontainers-1.14.3.jar:?]
	at com.github.dockerjava.core.exec.StartContainerCmdExec.execute(StartContainerCmdExec.java:31) ~[testcontainers-1.14.3.jar:?]
	at com.github.dockerjava.core.exec.StartContainerCmdExec.execute(StartContainerCmdExec.java:13) ~[testcontainers-1.14.3.jar:?]
	at com.github.dockerjava.core.exec.AbstrSyncDockerCmdExec.exec(AbstrSyncDockerCmdExec.java:21) ~[testcontainers-1.14.3.jar:?]
	at com.github.dockerjava.core.command.AbstrDockerCmd.exec(AbstrDockerCmd.java:35) ~[testcontainers-1.14.3.jar:?]
	at com.github.dockerjava.core.command.StartContainerCmdImpl.exec(StartContainerCmdImpl.java:43) ~[testcontainers-1.14.3.jar:?]
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:?]
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:?]
	at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
	at java.lang.reflect.Method.invoke(Method.java:566) ~[?:?]
	at org.testcontainers.dockerclient.AuditLoggingDockerClient.lambda$wrappedCommand$14(AuditLoggingDockerClient.java:102) ~[testcontainers-1.14.3.jar:?]
	at com.sun.proxy.$Proxy45.exec(Unknown Source) ~[?:?]
	at org.testcontainers.containers.GenericContainer.tryStart(GenericContainer.java:419) ~[testcontainers-1.14.3.jar:?]
	at org.testcontainers.containers.GenericContainer.lambda$doStart$0(GenericContainer.java:325) ~[testcontainers-1.14.3.jar:?]
	at org.rnorth.ducttape.unreliables.Unreliables.retryUntilSuccess(Unreliables.java:81) [duct-tape-1.0.8.jar:?]
	at org.testcontainers.containers.GenericContainer.doStart(GenericContainer.java:323) [testcontainers-1.14.3.jar:?]
	at org.testcontainers.containers.GenericContainer.start(GenericContainer.java:311) [testcontainers-1.14.3.jar:?]
	at com.hazelcast.jet.cdc.mysql.MySqlCdcNetworkIntegrationTest.initMySql(MySqlCdcNetworkIntegrationTest.java:355) [test-classes/:?]
	at com.hazelcast.jet.cdc.mysql.MySqlCdcNetworkIntegrationTest.when_databaseShutdownDuringBinlogReading(MySqlCdcNetworkIntegrationTest.java:258) [test-classes/:?]
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:?]
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:?]
	at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
	at java.lang.reflect.Method.invoke(Method.java:566) ~[?:?]
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50) [junit-4.12.jar:4.12]
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12) [junit-4.12.jar:4.12]
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47) [junit-4.12.jar:4.12]
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17) [junit-4.12.jar:4.12]
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:114) [hazelcast-4.0.3-SNAPSHOT-tests.jar:4.0.3-SNAPSHOT]
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:106) [hazelcast-4.0.3-SNAPSHOT-tests.jar:4.0.3-SNAPSHOT]
	at java.util.concurrent.FutureTask.run(FutureTask.java:264) [?:?]
	at java.lang.Thread.run(Thread.java:834) [?:?]
2020-09-20 13:40:34,883 [ERROR] [when_databaseShutdownDuringBinlogReading] [d.2]]: There are no stdout/stderr logs available for the failed container
2020-09-20 13:40:34,883 [ INFO] [when_databaseShutdownDuringBinlogReading] [d.2]]: Creating container for image: debezium/example-mysql:1.2
2020-09-20 13:40:35,414 [ INFO] [when_databaseShutdownDuringBinlogReading] [d.2]]: Starting container with ID: 36042f6571667c909788f0550aa03435a5214e6a6c71ad922cf0859ba5f1dc4f
2020-09-20 13:40:35,607 [ERROR] [when_databaseShutdownDuringBinlogReading] [d.2]]: Could not start container
com.github.dockerjava.api.exception.InternalServerErrorException: {"message":"driver failed programming external connectivity on endpoint infallible_elbakyan (ee951ee7d11957fb1fe03ca618104172637732b1e83e97617c9ed0fa774eb6f6): Error starting userland proxy: listen tcp 0.0.0.0:40708: bind: address already in use"}

	at com.github.dockerjava.okhttp.OkHttpInvocationBuilder.execute(OkHttpInvocationBuilder.java:293) ~[testcontainers-1.14.3.jar:?]
	at com.github.dockerjava.okhttp.OkHttpInvocationBuilder.execute(OkHttpInvocationBuilder.java:271) ~[testcontainers-1.14.3.jar:?]
	at com.github.dockerjava.okhttp.OkHttpInvocationBuilder.post(OkHttpInvocationBuilder.java:116) ~[testcontainers-1.14.3.jar:?]
	at com.github.dockerjava.core.exec.StartContainerCmdExec.execute(StartContainerCmdExec.java:31) ~[testcontainers-1.14.3.jar:?]
	at com.github.dockerjava.core.exec.StartContainerCmdExec.execute(StartContainerCmdExec.java:13) ~[testcontainers-1.14.3.jar:?]
	at com.github.dockerjava.core.exec.AbstrSyncDockerCmdExec.exec(AbstrSyncDockerCmdExec.java:21) ~[testcontainers-1.14.3.jar:?]
	at com.github.dockerjava.core.command.AbstrDockerCmd.exec(AbstrDockerCmd.java:35) ~[testcontainers-1.14.3.jar:?]
	at com.github.dockerjava.core.command.StartContainerCmdImpl.exec(StartContainerCmdImpl.java:43) ~[testcontainers-1.14.3.jar:?]
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:?]
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:?]
	at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
	at java.lang.reflect.Method.invoke(Method.java:566) ~[?:?]
	at org.testcontainers.dockerclient.AuditLoggingDockerClient.lambda$wrappedCommand$14(AuditLoggingDockerClient.java:102) ~[testcontainers-1.14.3.jar:?]
	at com.sun.proxy.$Proxy45.exec(Unknown Source) ~[?:?]
	at org.testcontainers.containers.GenericContainer.tryStart(GenericContainer.java:419) ~[testcontainers-1.14.3.jar:?]
	at org.testcontainers.containers.GenericContainer.lambda$doStart$0(GenericContainer.java:325) ~[testcontainers-1.14.3.jar:?]
	at org.rnorth.ducttape.unreliables.Unreliables.retryUntilSuccess(Unreliables.java:81) [duct-tape-1.0.8.jar:?]
	at org.testcontainers.containers.GenericContainer.doStart(GenericContainer.java:323) [testcontainers-1.14.3.jar:?]
	at org.testcontainers.containers.GenericContainer.start(GenericContainer.java:311) [testcontainers-1.14.3.jar:?]
	at com.hazelcast.jet.cdc.mysql.MySqlCdcNetworkIntegrationTest.initMySql(MySqlCdcNetworkIntegrationTest.java:355) [test-classes/:?]
	at com.hazelcast.jet.cdc.mysql.MySqlCdcNetworkIntegrationTest.when_databaseShutdownDuringBinlogReading(MySqlCdcNetworkIntegrationTest.java:258) [test-classes/:?]
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:?]
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:?]
	at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
	at java.lang.reflect.Method.invoke(Method.java:566) ~[?:?]
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50) [junit-4.12.jar:4.12]
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12) [junit-4.12.jar:4.12]
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47) [junit-4.12.jar:4.12]
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17) [junit-4.12.jar:4.12]
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:114) [hazelcast-4.0.3-SNAPSHOT-tests.jar:4.0.3-SNAPSHOT]
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:106) [hazelcast-4.0.3-SNAPSHOT-tests.jar:4.0.3-SNAPSHOT]
	at java.util.concurrent.FutureTask.run(FutureTask.java:264) [?:?]
	at java.lang.Thread.run(Thread.java:834) [?:?]
...
Hiccups measured while running test 'when_databaseShutdownDuringBinlogReading[fail](com.hazelcast.jet.cdc.mysql.MySqlCdcNetworkIntegrationTest):'
13:40:30, accumulated pauses: 51 ms, max pause: 20 ms, pauses over 1000 ms: 0
13:40:35, accumulated pauses: 14 ms, max pause: 6 ms, pauses over 1000 ms: 0
```

--- a/extensions/cdc-mysql/pom.xml
+++ b/extensions/cdc-mysql/pom.xml
@@ -132,6 +132,13 @@
             <version>${log4j2.slf4j.binding.version}</version>
             <scope>test</scope>
         </dependency>
+
+        <dependency>
+            <groupId>com.github.stefanbirkner</groupId>
+            <artifactId>system-rules</artifactId>
+            <version>1.19.0</version>
+            <scope>test</scope>
+        </dependency>
     </dependencies>
 
 </project>

--- a/extensions/cdc-mysql/pom.xml
+++ b/extensions/cdc-mysql/pom.xml
@@ -138,6 +138,12 @@
             <artifactId>system-rules</artifactId>
             <version>1.19.0</version>
             <scope>test</scope>
+            <exclusions>
+                <exclusion>
+                    <groupId>junit</groupId>
+                    <artifactId>junit-dep</artifactId>
+                </exclusion>
+            </exclusions>
         </dependency>
     </dependencies>
 

--- a/extensions/cdc-mysql/src/test/java/com/hazelcast/jet/cdc/mysql/MySqlCdcNetworkIntegrationTest.java
+++ b/extensions/cdc-mysql/src/test/java/com/hazelcast/jet/cdc/mysql/MySqlCdcNetworkIntegrationTest.java
@@ -101,6 +101,10 @@ public class MySqlCdcNetworkIntegrationTest extends AbstractCdcIntegrationTest {
 
     @Before
     public void before() {
+        //disable Testcontainer's automatic resource manager
+        //containers are cleaned up explicitly
+        //automatic resource manager is just an extra thing that can break
+        //(have had problems with it not being cleaned up properly itself)
         environmentVariables.set("TESTCONTAINERS_RYUK_DISABLED", "true");
         assertEquals("true", System.getenv("TESTCONTAINERS_RYUK_DISABLED"));
     }

--- a/extensions/cdc-mysql/src/test/java/com/hazelcast/jet/cdc/mysql/MySqlCdcNetworkIntegrationTest.java
+++ b/extensions/cdc-mysql/src/test/java/com/hazelcast/jet/cdc/mysql/MySqlCdcNetworkIntegrationTest.java
@@ -33,7 +33,10 @@ import com.hazelcast.jet.retry.RetryStrategy;
 import com.hazelcast.jet.test.SerialTest;
 import com.hazelcast.test.HazelcastSerialParametersRunnerFactory;
 import com.hazelcast.test.annotation.NightlyTest;
+import org.junit.Before;
+import org.junit.Rule;
 import org.junit.Test;
+import org.junit.contrib.java.lang.system.EnvironmentVariables;
 import org.junit.experimental.categories.Category;
 import org.junit.runner.RunWith;
 import org.junit.runners.Parameterized;
@@ -75,6 +78,9 @@ public class MySqlCdcNetworkIntegrationTest extends AbstractCdcIntegrationTest {
 
     private static final long RECONNECT_INTERVAL_MS = SECONDS.toMillis(1);
 
+    @Rule
+    public final EnvironmentVariables environmentVariables = new EnvironmentVariables();
+
     @Parameter(value = 0)
     public RetryStrategy reconnectBehavior;
 
@@ -91,6 +97,12 @@ public class MySqlCdcNetworkIntegrationTest extends AbstractCdcIntegrationTest {
                 { RetryStrategies.indefinitely(RECONNECT_INTERVAL_MS), false, "reconnect"},
                 { RetryStrategies.indefinitely(RECONNECT_INTERVAL_MS), true, "reconnect w/ state reset"}
         });
+    }
+
+    @Before
+    public void before() {
+        environmentVariables.set("TESTCONTAINERS_RYUK_DISABLED", "true");
+        assertEquals("true", System.getenv("TESTCONTAINERS_RYUK_DISABLED"));
     }
 
     @Test
@@ -139,31 +151,34 @@ public class MySqlCdcNetworkIntegrationTest extends AbstractCdcIntegrationTest {
                 ToxiproxyContainer toxiproxy = initToxiproxy(network);
         ) {
             MySQLContainer<?> mysql = initMySql(network, null);
-            ToxiproxyContainer.ContainerProxy proxy = initProxy(toxiproxy, mysql);
-            Pipeline pipeline = initPipeline(proxy.getContainerIpAddress(), proxy.getProxyPort());
-            // when job starts
-            JetInstance jet = createJetMembers(2)[0];
-            Job job = jet.newJob(pipeline);
-            assertJobStatusEventually(job, RUNNING);
-
-            // and snapshotting is ongoing (we have no exact way of identifying
-            // the moment, but random sleep will catch it at least some of the time)
-            TimeUnit.MILLISECONDS.sleep(ThreadLocalRandom.current().nextInt(0, 500));
-
-            // and connection is cut
-            proxy.setConnectionCut(true);
-
-            // and some time passes
-            SECONDS.sleep(2 * MILLISECONDS.toSeconds(RECONNECT_INTERVAL_MS));
-
-            // and connection recovers
-            proxy.setConnectionCut(false);
-
-            // then connector manages to reconnect and finish snapshot
             try {
-                assertEqualsEventually(() -> jet.getMap("results").size(), 4);
+                ToxiproxyContainer.ContainerProxy proxy = initProxy(toxiproxy, mysql);
+                Pipeline pipeline = initPipeline(proxy.getContainerIpAddress(), proxy.getProxyPort());
+                // when job starts
+                JetInstance jet = createJetMembers(2)[0];
+                Job job = jet.newJob(pipeline);
+                assertJobStatusEventually(job, RUNNING);
+
+                // and snapshotting is ongoing (we have no exact way of identifying
+                // the moment, but random sleep will catch it at least some of the time)
+                TimeUnit.MILLISECONDS.sleep(ThreadLocalRandom.current().nextInt(0, 500));
+
+                // and connection is cut
+                proxy.setConnectionCut(true);
+
+                // and some time passes
+                SECONDS.sleep(2 * MILLISECONDS.toSeconds(RECONNECT_INTERVAL_MS));
+
+                // and connection recovers
+                proxy.setConnectionCut(false);
+
+                // then connector manages to reconnect and finish snapshot
+                try {
+                    assertEqualsEventually(() -> jet.getMap("results").size(), 4);
+                } finally {
+                    abortJob(job);
+                }
             } finally {
-                abortJob(job);
                 stopContainer(mysql);
             }
         }
@@ -220,33 +235,36 @@ public class MySqlCdcNetworkIntegrationTest extends AbstractCdcIntegrationTest {
                 ToxiproxyContainer toxiproxy = initToxiproxy(network);
         ) {
             MySQLContainer<?> mysql = initMySql(network, null);
-            ToxiproxyContainer.ContainerProxy proxy = initProxy(toxiproxy, mysql);
-            Pipeline pipeline = initPipeline(proxy.getContainerIpAddress(), proxy.getProxyPort());
-            // when connector is up and transitions to binlog reading
-            JetInstance jet = createJetMembers(2)[0];
-            Job job = jet.newJob(pipeline);
-            assertEqualsEventually(() -> jet.getMap("results").size(), 4);
-            SECONDS.sleep(3);
-            insertRecords(mysql, 1005);
-            assertEqualsEventually(() -> jet.getMap("results").size(), 5);
-
-            // and the connection is cut
-            proxy.setConnectionCut(true);
-
-            // and some new events get generated in the DB
-            insertRecords(mysql, 1006, 1007);
-
-            // and some time passes
-            TimeUnit.MILLISECONDS.sleep(2 * RECONNECT_INTERVAL_MS);
-
-            // and the connection is re-established
-            proxy.setConnectionCut(false);
-
-            // then the connector catches up
             try {
-                assertEqualsEventually(() -> jet.getMap("results").size(), 7);
+                ToxiproxyContainer.ContainerProxy proxy = initProxy(toxiproxy, mysql);
+                Pipeline pipeline = initPipeline(proxy.getContainerIpAddress(), proxy.getProxyPort());
+                // when connector is up and transitions to binlog reading
+                JetInstance jet = createJetMembers(2)[0];
+                Job job = jet.newJob(pipeline);
+                assertEqualsEventually(() -> jet.getMap("results").size(), 4);
+                SECONDS.sleep(3);
+                insertRecords(mysql, 1005);
+                assertEqualsEventually(() -> jet.getMap("results").size(), 5);
+
+                // and the connection is cut
+                proxy.setConnectionCut(true);
+
+                // and some new events get generated in the DB
+                insertRecords(mysql, 1006, 1007);
+
+                // and some time passes
+                TimeUnit.MILLISECONDS.sleep(2 * RECONNECT_INTERVAL_MS);
+
+                // and the connection is re-established
+                proxy.setConnectionCut(false);
+
+                // then the connector catches up
+                try {
+                    assertEqualsEventually(() -> jet.getMap("results").size(), 7);
+                } finally {
+                    abortJob(job);
+                }
             } finally {
-                abortJob(job);
                 stopContainer(mysql);
             }
         }

--- a/extensions/cdc-postgres/pom.xml
+++ b/extensions/cdc-postgres/pom.xml
@@ -132,6 +132,13 @@
             <version>${log4j2.slf4j.binding.version}</version>
             <scope>test</scope>
         </dependency>
+
+        <dependency>
+            <groupId>com.github.stefanbirkner</groupId>
+            <artifactId>system-rules</artifactId>
+            <version>1.19.0</version>
+            <scope>test</scope>
+        </dependency>
     </dependencies>
 
 </project>

--- a/extensions/cdc-postgres/pom.xml
+++ b/extensions/cdc-postgres/pom.xml
@@ -138,6 +138,12 @@
             <artifactId>system-rules</artifactId>
             <version>1.19.0</version>
             <scope>test</scope>
+            <exclusions>
+                <exclusion>
+                    <groupId>junit</groupId>
+                    <artifactId>junit-dep</artifactId>
+                </exclusion>
+            </exclusions>
         </dependency>
     </dependencies>
 

--- a/extensions/cdc-postgres/src/test/java/com/hazelcast/jet/cdc/postgres/PostgresCdcNetworkIntegrationTest.java
+++ b/extensions/cdc-postgres/src/test/java/com/hazelcast/jet/cdc/postgres/PostgresCdcNetworkIntegrationTest.java
@@ -34,7 +34,10 @@ import com.hazelcast.jet.test.SerialTest;
 import com.hazelcast.test.HazelcastSerialParametersRunnerFactory;
 import com.hazelcast.test.annotation.NightlyTest;
 import org.junit.Assume;
+import org.junit.Before;
+import org.junit.Rule;
 import org.junit.Test;
+import org.junit.contrib.java.lang.system.EnvironmentVariables;
 import org.junit.experimental.categories.Category;
 import org.junit.runner.RunWith;
 import org.junit.runners.Parameterized;
@@ -76,6 +79,9 @@ public class PostgresCdcNetworkIntegrationTest extends AbstractCdcIntegrationTes
 
     private static final long RECONNECT_INTERVAL_MS = SECONDS.toMillis(1);
 
+    @Rule
+    public final EnvironmentVariables environmentVariables = new EnvironmentVariables();
+
     @Parameter(value = 0)
     public RetryStrategy reconnectBehavior;
 
@@ -92,6 +98,12 @@ public class PostgresCdcNetworkIntegrationTest extends AbstractCdcIntegrationTes
                 { RetryStrategies.indefinitely(RECONNECT_INTERVAL_MS), false, "reconnect"},
                 { RetryStrategies.indefinitely(RECONNECT_INTERVAL_MS), true, "reconnect w/ state reset"}
         });
+    }
+
+    @Before
+    public void before() {
+        environmentVariables.set("TESTCONTAINERS_RYUK_DISABLED", "true");
+        assertEquals("true", System.getenv("TESTCONTAINERS_RYUK_DISABLED"));
     }
 
     @Test
@@ -139,32 +151,35 @@ public class PostgresCdcNetworkIntegrationTest extends AbstractCdcIntegrationTes
                 ToxiproxyContainer toxiproxy = initToxiproxy(network);
         ) {
             PostgreSQLContainer<?> postgres = initPostgres(network, null);
-            ToxiproxyContainer.ContainerProxy proxy = initProxy(toxiproxy, postgres);
-            Pipeline pipeline = initPipeline(proxy.getContainerIpAddress(), proxy.getProxyPort());
-            // when job starts
-            JetInstance jet = createJetMembers(2)[0];
-            Job job = jet.newJob(pipeline);
-            assertJobStatusEventually(job, RUNNING);
-
-            // and snapshotting is ongoing (we have no exact way of identifying
-            // the moment, but random sleep will catch it at least some of the time)
-            MILLISECONDS.sleep(ThreadLocalRandom.current().nextInt(0, 500));
-
-            // and connection is cut
-            proxy.setConnectionCut(true);
-
-            // and some time passes
-            MILLISECONDS.sleep(2 * RECONNECT_INTERVAL_MS);
-            //it takes the bloody thing 150 seconds to notice the connection being down, so it won't notice this...
-
-            // and connection recovers
-            proxy.setConnectionCut(false);
-
-            // then connector manages to reconnect and finish snapshot
             try {
-                assertEqualsEventually(() -> jet.getMap("results").size(), 4);
+                ToxiproxyContainer.ContainerProxy proxy = initProxy(toxiproxy, postgres);
+                Pipeline pipeline = initPipeline(proxy.getContainerIpAddress(), proxy.getProxyPort());
+                // when job starts
+                JetInstance jet = createJetMembers(2)[0];
+                Job job = jet.newJob(pipeline);
+                assertJobStatusEventually(job, RUNNING);
+
+                // and snapshotting is ongoing (we have no exact way of identifying
+                // the moment, but random sleep will catch it at least some of the time)
+                MILLISECONDS.sleep(ThreadLocalRandom.current().nextInt(0, 500));
+
+                // and connection is cut
+                proxy.setConnectionCut(true);
+
+                // and some time passes
+                MILLISECONDS.sleep(2 * RECONNECT_INTERVAL_MS);
+                //it takes the bloody thing 150 seconds to notice the connection being down, so it won't notice this...
+
+                // and connection recovers
+                proxy.setConnectionCut(false);
+
+                // then connector manages to reconnect and finish snapshot
+                try {
+                    assertEqualsEventually(() -> jet.getMap("results").size(), 4);
+                } finally {
+                    abortJob(job);
+                }
             } finally {
-                abortJob(job);
                 stopContainer(postgres);
             }
         }
@@ -222,35 +237,38 @@ public class PostgresCdcNetworkIntegrationTest extends AbstractCdcIntegrationTes
                 ToxiproxyContainer toxiproxy = initToxiproxy(network);
         ) {
             PostgreSQLContainer<?> postgres = initPostgres(network, null);
-            ToxiproxyContainer.ContainerProxy proxy = initProxy(toxiproxy, postgres);
-            Pipeline pipeline = initPipeline(proxy.getContainerIpAddress(), proxy.getProxyPort());
-            // when connector is up and transitions to binlog reading
-            JetInstance jet = createJetMembers(2)[0];
-            Job job = jet.newJob(pipeline);
-            assertEqualsEventually(() -> jet.getMap("results").size(), 4);
-            SECONDS.sleep(3);
-            insertRecords(postgres, 1005);
-            assertEqualsEventually(() -> jet.getMap("results").size(), 5);
-
-            // and the connection is cut
-            proxy.setConnectionCut(true);
-
-            // and some new events get generated in the DB
-            insertRecords(postgres, 1006, 1007);
-
-            // and some time passes
-            MILLISECONDS.sleep(5 * RECONNECT_INTERVAL_MS);
-
-            // and the connection is re-established
-            proxy.setConnectionCut(false);
-
-            // then
             try {
-                // then job keeps running, connector starts freshly, including snapshotting
-                assertEqualsEventually(() -> jet.getMap("results").size(), 7);
-                assertEquals(RUNNING, job.getStatus());
+                ToxiproxyContainer.ContainerProxy proxy = initProxy(toxiproxy, postgres);
+                Pipeline pipeline = initPipeline(proxy.getContainerIpAddress(), proxy.getProxyPort());
+                // when connector is up and transitions to binlog reading
+                JetInstance jet = createJetMembers(2)[0];
+                Job job = jet.newJob(pipeline);
+                assertEqualsEventually(() -> jet.getMap("results").size(), 4);
+                SECONDS.sleep(3);
+                insertRecords(postgres, 1005);
+                assertEqualsEventually(() -> jet.getMap("results").size(), 5);
+
+                // and the connection is cut
+                proxy.setConnectionCut(true);
+
+                // and some new events get generated in the DB
+                insertRecords(postgres, 1006, 1007);
+
+                // and some time passes
+                MILLISECONDS.sleep(5 * RECONNECT_INTERVAL_MS);
+
+                // and the connection is re-established
+                proxy.setConnectionCut(false);
+
+                // then
+                try {
+                    // then job keeps running, connector starts freshly, including snapshotting
+                    assertEqualsEventually(() -> jet.getMap("results").size(), 7);
+                    assertEquals(RUNNING, job.getStatus());
+                } finally {
+                    abortJob(job);
+                }
             } finally {
-                abortJob(job);
                 stopContainer(postgres);
             }
         }

--- a/extensions/cdc-postgres/src/test/java/com/hazelcast/jet/cdc/postgres/PostgresCdcNetworkIntegrationTest.java
+++ b/extensions/cdc-postgres/src/test/java/com/hazelcast/jet/cdc/postgres/PostgresCdcNetworkIntegrationTest.java
@@ -102,6 +102,10 @@ public class PostgresCdcNetworkIntegrationTest extends AbstractCdcIntegrationTes
 
     @Before
     public void before() {
+        //disable Testcontainer's automatic resource manager
+        //containers are cleaned up explicitly
+        //automatic resource manager is just an extra thing that can break
+        //(have had problems with it not being cleaned up properly itself)
         environmentVariables.set("TESTCONTAINERS_RYUK_DISABLED", "true");
         assertEquals("true", System.getenv("TESTCONTAINERS_RYUK_DISABLED"));
     }

